src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs

```diff
@@ -317,32 +449,33 @@ static class PopupPositionerExtensions
                 if (matrix == null)
                 {
                     if (target.GetVisualRoot() == null)
-                        throw new InvalidCastException("Target control is not attached to the visual tree");
-                    throw new InvalidCastException("Target control is not in the same tree as the popup parent");
+                        throw new InvalidOperationException("Target control is not attached to the visual tree");
+                    throw new InvalidOperationException("Target control is not in the same tree as the popup parent");
                 }
 
-                positionerParameters.AnchorRectangle = new Rect(default, target.Bounds.Size)
-                    .TransformToAABB(matrix.Value);
+                var bounds = new Rect(default, target.Bounds.Size);
+                var anchorRect = rect ?? bounds;
+                positionerParameters.AnchorRectangle = anchorRect.Intersect(bounds).TransformToAABB(matrix.Value);
```

Comment:
@kekekeks I added an intersection with `bounds` here because from what I understand the anchor rectangle needs to be within the control - is this correct?

Reply:
We need to check how WPF handles the placement rect, but for me it makes sense to intersect with the control bounds, yes.

Reply:
Actually, there might be a problem when placement rect is somehow outside of the control area. We might end up with a 0x0 rect which is invalid.

Reply:
Well a control's bounds may also be of size 0,0... Maybe we should treat a 0-sized rect as a one-sized rect?